Demonstrate enhanced-parallelism Dormand-Prince stepper

simbarras opened the pull request:
This is my first attempt to optimize Runge Kutta Dormand Prince using four threads to compute a chord. This implementation is using a controller thread that computes the serial part using the equation and three workers threads that update the states of the particle. The threads are synchronized using "__syncwarp()" and the pointers are located on the global memory.

![result](https://github.com/celeritas-project/celeritas/assets/66463606/c8e7bf6f-0f17-46ed-9e22-b6801c7ad8a2)

The graph shows an improvement from 35% to 40%. The difference between v0.1 and v0.2 is when the coefficients are computed, and this pull request is on top of v0.1.

To know the speed, a unit test can be used. I personally run the test from the celeritas directory:
``` bash
./scripts/build.sh base --no-test && build/test/celeritas_field_DormandPrinceStepper
```

related: https://github.com/simbarras/tb23-gpu-opt-celeritas/issues/22

Commit:
Add test that compute the time to complete a kernel
Add option to skip the build or the test in the celeritas script

Commit:
Update test framework

Commit:
New implementation that use aside thread to compute vector multiplication

Commit:
Reformat files

sethrj reviewed (changes requested):
@simbarras The change files show a bunch of superfluous changes to the CMakeLists and h.in files: maybe line endings? (Are you developing on a windows system?) Please check those and revert them. Also, since the new CUDA code has to be inlined and available elsewhere, please combine the multistepper files into a single `.cuda.hh` file for inclusion elsewhere. Thanks!

Commit:
Fix comment made on the PR.
Fix end of line character
Convert multi stepper into a .cuda.hh file to allow inlining

Commit:
Run two implementation to compare them
The difference between the two is the usage or not of the shared memory

Commit:
Compare the two implementation

simbarras commented:
I have done some experiments to find the limit of the two implementations I have made.

![image](https://github.com/celeritas-project/celeritas/assets/66463606/1c564a00-1c16-48f4-9199-850c3d3748dd)

The limit reached with the shared memory implementation is logical and can be calculated by knowing the size of the objects in front of it.
```
45 * ( 7 + 1) + 144 = 528 bytes/track
^       ^    ^       ^
|       |    |       |----- FieldStepperResult's size
|       |    |------------- State updated along the whole method
|       | ----------------- K's states
|-------------------------- OdeState's size
 ```


Concerning the error that appeared when implementing global memory, the only thing that is indicated is an excess in the requested resources, but we don't know which one.

```
CUDA error: too many resources requested for launch
Launc config for 193 states: 1 blocks, 772 threads per block, 0 bytes of shared memory and 111176 bytes of arguments
Device properties: 1024 threads per block, 2147483647 blocks and 49152 bytes of shared memory
```

For the classic implementation error, it's a failed assertion.
```
.../celeritas/src/celeritas/field/MagFieldEquation.hh:101: celeritas::OdeState celeritas::MagFieldEquation<FieldT>::operator()(const celeritas::OdeState &) const [with FieldT = celeritas::Array<double, 3U> (&)(const celeritas::Array<double, 3U> &)]: block: [0,0,0], thread: [427,0,0] Assertion `false && sizeof("precondition" ": " "y.mom[0] != 0 || y.mom[1] != 0 || y.mom[2] != 0")` failed.
```

sethrj commented:
@simbarras the block size `2147483647` is definitely an overflow...

esseivaju commented on src/celeritas/field/DormandPrinceMultiStepperGlobal.cuda.hh:
```diff
@@ -0,0 +1,280 @@
+                                          OdeState* along_state,
+                                          FieldStepperResult* result) const;
+
+    CELER_FUNCTION void run_sequential(real_type step,
```
`run_sequential` and `run_aside` can probably be private.

esseivaju commented on src/celeritas/field/DormandPrinceMultiStepperGlobal.cuda.hh:
```diff
@@ -0,0 +1,280 @@
+//---------------------------------------------------------------------------//
+// MACROS
+//---------------------------------------------------------------------------//
+#define UPDATE_STATE(index, state, coefficient, k)        \
```
undef macros once you no longer need them (at the end of the header). Also if you can use lambdas instead.

esseivaju commented on src/celeritas/field/DormandPrinceMultiStepperGlobal.cuda.hh:
```diff
@@ -0,0 +1,280 @@
+//---------------------------------------------------------------------------//
+#if !CELER_USE_DEVICE
+
+// TODO: implement this for CPU
```
Is there even a CPU implementation that makes sense here? I'm not sure it's worth trying to implement the same algorithm for CPU since threads are limited.

esseivaju commented on src/celeritas/field/DormandPrinceMultiStepperGlobal.cuda.hh:
```diff
@@ -0,0 +1,280 @@
+    int id = (threadIdx.x + blockIdx.x * blockDim.x) / number_threads;
+    int index = (threadIdx.x + blockIdx.x * blockDim.x) % number_threads;
+
+    int mask = (4 * 4 - 1) << ((id * 4) % 32);
```
maybe have constexpr naming these numbers?

esseivaju commented on test/celeritas/field/DormandPrinceStepper.test.cu:
```diff
@@ -0,0 +1,318 @@
+    // Allocate memory on device
+    cudaMalloc(&d_results, number_states * sizeof(FieldStepperResult));
+    cudaMalloc(&d_states, number_states * sizeof(OdeState));
+    cudaMalloc(&d_number_iterations, sizeof(int));
```
you don't really need to allocate small (less than pointer size?) variables to global memory, it's faster and clearer to let CUDA do a copy and have the variable in register.

esseivaju commented on test/celeritas/field/DormandPrinceStepper.test.hh:
```diff
@@ -0,0 +1,227 @@
+constexpr int one_thread = 1;
+constexpr int multi_thread = 4;
+constexpr int number_iterations = 40;
+constexpr OdeState initial_states[5] = {
```
It would be interesting to see measurement with more tracks to fill the GPU

simbarras replied:
I use these five states to generate an array of values with a dynamic length ([code here](https://github.com/simbarras/celeritas/blob/7bf84b821c22ba60f9b60ff11793656242a07546/test/celeritas/field/DormandPrinceStepper.test.cu#L158-L164)). The second graph shows the result of the experiment with different number of tracks.

simbarras commented:
> @simbarras the block size `2147483647` is definitely an overflow...
 In this log
 
Sorry for the confusion. Line `1` is the error reported by CUDA, line `2` is the configuration used to launch the kernel and line `3` is the maximum values accepted by the device.
 ```
1: CUDA error: too many resources requested for launch
2: Launch config for 193 states: 1 blocks, 772 threads per block, 0 bytes of shared memory and 111176 bytes of arguments
3: Device properties: 1024 threads per block, 2147483647 blocks and 49152 bytes of shared memory
```

sethrj commented on test/celeritas/field/DormandPrinceStepper.test.hh:
```diff
@@ -0,0 +1,227 @@
+    result += "Actual mid state position:     "
+              + std::to_string(actual.mid_state.pos[0]) + ", "
+              + std::to_string(actual.mid_state.pos[1]) + ", "
+              + std::to_string(actual.mid_state.pos[2]) + "\n";
```
You should be using the googletest `EXPECT_VEC_SOFT_EQ` and include `corecel/math/ArrayIO.hh` to have printable arrays.

esseivaju replied:
In addition you can wrap this in a lambda:

```suggestion
    auto format_results = [](std::string prefix, const Real3& r){
        return prefix
                 + std::to_string(r[0]) + ", "
                 + std::to_string(r[1]) + ", "
                 + std::to_string(r[2]) + "\n";
    };
    std::string result = format_results("Expected mid state position:   ", expected.mid_state.pos);
    result += format_results("Actual mid state position:     ", actual.mid_state.pos);
```

sethrj commented on test/celeritas/field/DormandPrinceStepper.test.hh:
```diff
@@ -0,0 +1,227 @@
+{
+//---------------------------------------------------------------------------//
+using celeritas::units::ElementaryCharge;
+using Evaluator_t = celeritas::MagFieldEquation<Real3 (&)(Real3 const&)>;
```
Instead of a device function pointer, use a function-like object (struct) and template on that. This is what we do in practice, and the indirection on such a tight loop will seriously affect performance.

sethrj commented on test/celeritas/field/DormandPrinceStepper.test.hh:
```diff
@@ -0,0 +1,227 @@
+using celeritas::units::ElementaryCharge;
+using Evaluator_t = celeritas::MagFieldEquation<Real3 (&)(Real3 const&)>;
+using Stepper_uni = celeritas::DormandPrinceStepper<Evaluator_t&>;
+using Stepper_multi_global
```
Use class-like names for type aliases
```suggestion
using StepperMultiGlobla
```

sethrj commented on test/celeritas/field/DormandPrinceStepper.test.cu:
```diff
@@ -0,0 +1,318 @@
+__global__ void dormand_test_arg_kernel(OdeState* states,
+                                        FieldStepperResult* results,
+                                        int* number_iterations,
+                                        int* number_threads,
```
There's need to pass these by pointer since they're input-only scalars : pass them by value instead.

sethrj commented on test/celeritas/field/DormandPrinceStepper.test.cu:
```diff
@@ -0,0 +1,318 @@
+    arguments_size += sizeof(int) * 2;
+
+    FieldStepperResult *h_results, *d_results;
+    h_results = new FieldStepperResult[number_states];
```
Use `std::vector` (or `thrust` host vector) instead of `new`. Use thrust device vector or `corecel/data/DeviceVector.hh`.

sethrj commented on test/celeritas/field/DormandPrinceStepper.test.cu:
```diff
@@ -0,0 +1,318 @@
+            = detail::distance_chord(state, res.mid_state, res.end_state);
+        step *= max(std::sqrt(delta_chord / dchord), half);
+    }
+    results[id] = res;
```
It's possible some of your errors are due to invalid memory accesses. It's good practice to pass the vector size as a const kernel argument and add an assertion `CELER_ASSERT(id < results_size);` before assigning.

sethrj commented on test/celeritas/field/DormandPrinceStepper.test.cu:
```diff
@@ -0,0 +1,318 @@
+                                          FieldStepperResult* result)
+{
+    return FieldStepperResult();
+}
```
If the generic template isn't used, simply declare it instead of making an invalid definition. I think you should actually be able to overload rather than specializing on the template though.

sethrj commented on test/celeritas/field/DormandPrinceStepper.test.cu:
```diff
@@ -0,0 +1,318 @@
+}
+
+template<>
+__device__ FieldStepperResult run_stepper(Stepper_uni stepper,
```
The stepper should be passed by reference, as should the input ODE state.

sethrj commented on test/celeritas/field/DormandPrinceStepper.test.cc:
```diff
@@ -0,0 +1,143 @@
+{
+    constexpr int number_max_states = 427;
+#define debug
+#ifdef debug
```
Avoid macros that comment out large blocks of code, as well as macros that have the name of common expressions. If you need to, make two separate tests and prefix one with `DISABLED_` to tell googletest not to run it (`compare_time`, `DISABLED_compare_time_all`).

sethrj reviewed (commented):
Does `111176 bytes of arguments` refer to the actual kernel argument size (which it doesn't look like from what I can tell), or the total global memory allocation? If the former, that could definitely make the CUDA compiler unhappy.

esseivaju commented on src/celeritas/field/DormandPrinceMultiStepperGlobal.cuda.hh:
```diff
@@ -0,0 +1,280 @@
+// MACROS
+//---------------------------------------------------------------------------//
+#define UPDATE_STATE(index, state, coefficient, k)        \
+    state.pos[index - 1] = coefficient * k.pos[index - 1] \
```
since you're only using `index - 1` I think it'd be more intuitive to use index in the macro and pass index - 1 as argument.

esseivaju commented on test/celeritas/field/DormandPrinceStepper.test.cu:
```diff
@@ -0,0 +1,318 @@
+        if (use_shared)
+        {
+            dormand_test_arg_kernel<Stepper_multi_shared>
+                <<<block_dimension, thread_dimension, shared_memory>>>(
```
`block_dimension` and `thread_dimension` are a bit confusing names. The first parameter is the grid size (i.e. # of blocks) and the second parameter is the block size.

esseivaju commented on test/celeritas/field/DormandPrinceStepper.test.hh:
```diff
@@ -0,0 +1,227 @@
+}
+
+inline CELER_FUNCTION bool
+compare_results(FieldStepperResult& e1, FieldStepperResult& e2)
```
you can probably use a loop to check these conditions, it's quite verbose.

esseivaju commented on src/celeritas/field/DormandPrinceMultiStepperShared.cuda.hh:
```diff
@@ -0,0 +1,296 @@
+    int id = (threadIdx.x + blockIdx.x * blockDim.x) / number_threads;
+    int index = (threadIdx.x + blockIdx.x * blockDim.x) % number_threads;
+
+    extern __shared__ void* shared_memory[];
+    OdeState* shared_ks = (OdeState*)shared_memory;
+    OdeState* shared_along_state
+        = reinterpret_cast<OdeState*>(&shared_ks[7 * num_states]);
+    FieldStepperResult* shared_result = reinterpret_cast<FieldStepperResult*>(
+        &shared_along_state[num_states]);
```
You might want to add assertion to make sure that shared memory is getting indexed correctly.

esseivaju replied:
Also, are you sure that shared_result always has the correct memory alignment?

esseivaju commented on test/celeritas/field/DormandPrinceStepper.test.cu:
```diff
@@ -0,0 +1,318 @@
+    constexpr double delta_chord = 1e-4;
+    constexpr double half = 0.5;
+
+    auto id = (blockIdx.x * blockDim.x + threadIdx.x) / *number_threads;
```
you can use `KernelParamCalculator::thread_id()` instead of using builtin CUDA variables.

esseivaju commented on test/celeritas/field/DormandPrinceStepper.test.cu:
```diff
@@ -0,0 +1,318 @@
+// TESTING INTERFACE
+//---------------------------------------------------------------------------//
+//! Run on device and return results
+KernelResult simulate_multi_next_chord(int number_threads,
```
this function is very long, I'd try to break it down to make it more readable.

Commit:
Fix most of the comments of Seth and Julien

esseivaju commented on test/celeritas/field/DormandPrinceStepper.test.cu:
```diff
@@ -0,0 +1,318 @@
+                        + number_states * sizeof(FieldStepperResult);
+    }
+    int block_dimension = 1;
+    int thread_dimension = number_threads * number_states;
```
maximum across CUDA architecture is 1024, in practice probably less because of registers usage. Maybe have a fixed block size and vary the number of blocks? It also makes it easier for the shared memory case; it won't grow out of control over the physical device limit.

esseivaju commented on test/celeritas/field/DormandPrinceStepper.test.hh:
```diff
@@ -0,0 +1,227 @@
+
+//---------------------------------------------------------------------------//
+//! Run on device and return results
+KernelResult simulate_multi_next_chord(int number_threads,
+                                       int number_states,
+                                       bool use_shared = false);
```
```suggestion
KernelResult simulate_multi_next_chord(int number_threads,
                                       int number_states,
                                       bool use_shared);
```

sethrj commented on src/celeritas/field/DormandPrinceMultiStepperShared.cuda.hh:
```diff
@@ -0,0 +1,363 @@
+inline CELER_FUNCTION void DormandPrinceMultiStepperShared<E>::update_state(
+    int index, OdeState& state, real_type coefficient, OdeState const& k) const
+{
+    state.pos[index - 1] = coefficient * k.pos[index - 1]
+                           + state.pos[index - 1];
+    state.mom[index - 1] = coefficient * k.mom[index - 1]
+                           + state.mom[index - 1];
```
Since these two calculations are independent, couldn't you double the parallelism by having two separate threads perform this calculation too?

simbarras replied:
It is an idee but we see that the requested resources are already too high, I do not think this will speed up the process.

Commit:
WIP: Allow kernel to be launch with multiple block

Commit:
WIP: refactor simulate_multi_next_chord to do it more readable

Commit:
WIP: refactor compute of the kernel configuration

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

sethrj committed:
Merge branch 'develop' into develop
